### PR TITLE
New packages: rockpro64-base and rockpro64-uboot

### PR DESCRIPTION
**Pull request by tiemenwerkman**
#### General
- [X] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [X] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [ ] I generally don't use the affected packages but briefly tested this PR

#### Does it build and run successfully? 
(Please choose at least one native build and, if supported, at least one cross build. More are better.)
- [ ] I built this PR locally for my native architecture, (ARCH-LIBC)
- [ ] I built this PR locally for these architectures (if supported. mark crossbuilds):
  - [ ] aarch64-musl
  - [ ] armv7l
  - [ ] armv6l-musl

I have not built the package native on aarch64, only crossbuilds on x86_64 with aarch64 and aarch64-musl. Both built successfully with xbps-src -K.
Most of the work is copied from the pinebook-pro packages, as both are very similar using the RK3399 soc. I did not create a rockpro64 specific kernel as it did not seem necessary. I have tested using kernels 5.10, 5.13 and 5.14. The latter two versions produce fewer errors when booting (loglevel=4).
I don't know if this is relevant but I have also created images with the zfs dkms module. It only succeeded building on kernel version 5.10, however not without errors, zfs does work though.
I have not updated U-Boot to the latest version (2021-10 as of now) because it fails to build. Earlier this week it was fine so I don't yet know what happened. The second to last version (2021-07) is still very recent and does work fine.
The pinebook-pro template creates a boot.scr to boot the image, the rock64 template uses extlinux to boot, after a bit of research I opted to go with extlinux because that is what everyone else is doing (manjaro, ayufan, armbian) and this simplifies the patch for void-mklive to create an image, it does require u-boot-menu as an extra dependency.
Most of the patches come straight from the pinebookpro-uboot package with a few features taken from [Sigmaris's U-Boot](https://github.com/sigmaris/u-boot/). A description of the details can be found in the patch files.

### CODE REVIEW

**Comment by CameronNemo** [comment]
Linted, built on aarch64 and aarch64-musl, reviewed templates et al. LGTM. But I do not have the RockPro64 to test with.

@Anachron

**Comment by Anachron** [comment]
I'm going to buy a RockPro64 myself and test this PR... stay tuned.

**Comment by Anachron** [comment]
I was able to build this as expected:

`Created /home/anon/app/dat/mkvr/images/void-live-x86_64-20220317-rockpro64-de_DE.UTF-8.iso (560M)
successfully.`

I'm going to test this on my hardware soon and report once again

**Comment by Anachron** [comment]
Just a headsup:
It surely would be good to include https://github.com/tuxd3v/ats in the repo after this is merged because the fan is quite noisy by default.

It supports sysvinit so maybe we can patch it for runit. (I might even give it a try)

So far have not tried booting yet, I need to get a new microsd card before.

**Comment by Anachron** [comment]
@tiemenwerkman @tsndqst @CameronNemo please tell me on how to get a working `platformfs` for `rockpro64`.

What I did was I took the `void-aarch64-ROOTFS-20210930.tar.xz` and added the `rockpro64-base` to it:

```sh
./mkplatformfs.sh  -p rockpro64-base -c /home/anon/.mkvr/cache -r /home/anon/.bldr/host/binpkgs/pr-33556 -o /home/anon/.mkvr/images/void-platform-aarch64-20220318-rockpro64-de_DE.UTF-8.tar.xz rockpro64 /home/anon/.mkvr/data/rockpro64/rootfs.tar.xz
```

Surely I must be missing something here?

**Comment by tsndqst** [comment]
@Anachron I used

**Comment by Anachron** [comment]
Okay, thanks for the quick reply, I figure I need a `linux` package but wasn't sure if that really was the case because the `mklive` adds one by it self. Will recreate the rootfs and try again another day.

Also, what's inside your `/rp64_post_cmd`?

**Comment by tsndqst** [comment]
I recall being surprised that I had to add a `linux` package too.

My `rp64_post_cmd` script has some setup stuff that I got tired of running on my three rockpro64s after install.  Half of is it prepping for installing kubernetes.

```
#!/bin/bash
chroot $1 /bin/bash <<EOT

ln -sf /etc/sv/socklog-unix /etc/runit/runsvdir/current
ln -sf /etc/sv/nanoklogd /etc/runit/runsvdir/current 

echo "Set timezone"
ln -sf /usr/share/zoneinfo/US/Central /etc/localtime

echo "Start ntpd"
ntpd -gq
ln -s /etc/sv/ntpd /etc/runit/runsvdir/current

echo "Configure dhcpcd"
sed -i -e '/^\#hostname/s/\#//' -e '/option host_name/a env force_hostname=YES\nenv hostname_fqdn=false' /etc/dhcpcd.conf 
echo 'denyinterfaces cilium*,lxc*' >> /etc/dhcpcd.conf

echo "Set up salt"
sed -i '/^#master:/s/\#//' /etc/salt/minion
ln -s /etc/sv/salt-minion /etc/runit/runsvdir/current

echo "Set up containerd"
mkdir -p /etc/containerd
containerd config default > /etc/containerd/config.toml
ln -s /etc/sv/containerd /etc/runit/runsvdir/current

echo "Set up containerd modules"
cat <<EOF > /etc/modules-load.d/containerd.conf
overlay
br_netfilter
EOF

echo "Setup required sysctl params"
mkdir -p /etc/sysctl.d
cat <<EOF > /etc/sysctl.d/99-kubernetes-cri.conf
net.bridge.bridge-nf-call-iptables  = 1
net.ipv4.ip_forward                 = 1
net.bridge.bridge-nf-call-ip6tables = 1
EOF

echo "Create /etc/crictl.yaml"
echo "runtime-endpoint: unix:///run/containerd/containerd.sock" > /etc/crictl.yaml

echo CGROUP_MODE=unified >> /etc/rc.conf

echo "Mount / as rshared"
sed -i -e '/\//s/defaults/defaults,rshared/' /etc/fstab

mkdir -p /etc/kubernetes/manifests

#
# Update kubelet config.yaml to use cgroupfs cgroupDriver
#
cat <<EOF > /etc/kubernetes/kubelet-config.yaml
apiVersion: kubelet.config.k8s.io/v1beta1
cgroupDriver: cgroupfs
serverTLSBootstrap: true
EOF

sync
EOT
```

**Comment by Anachron** [comment]
More info.

Inside `Armbian/boot`:

```
Image:                          symbolic link to vmlinuz-5.15.25-rockchip64
System.map-5.15.25-rockchip64:  ASCII text
armbianEnv.txt:                 ASCII text
armbian_first_run.txt.template: ASCII text
boot.bmp:                       PC bitmap, Windows 3.x format, 320 x 240 x 4, image size 38400, cbSize 38518, bits offset 118
boot.cmd:                       ASCII text
boot.scr:                       u-boot legacy uImage, , Linux/ARM, Script File (Not compressed), 3121 bytes, Thu Aug 26 11:08:57 2021, Load Address: 00000000, Entry Point: 00000000, Header CRC: 0XBE5168B2, Data CRC: 0X7F0AB2DF
config-5.15.25-rockchip64:      Linux make config build file, ASCII text
dtb:                            symbolic link to dtb-5.15.25-rockchip64
dtb-5.15.25-rockchip64:         directory
initrd.img-5.10.60-rockchip64:  gzip compressed data, was "mkinitramfs-MAIN_6O9r8Y", last modified: Wed Mar 16 22:03:52 2022, from Unix, original size modulo 2^32 48922624
initrd.img-5.15.25-rockchip64:  gzip compressed data, was "mkinitramfs-MAIN_4e0IWN", last modified: Thu Mar 17 20:47:05 2022, from Unix, original size modulo 2^32 28397056
uInitrd:                        symbolic link to uInitrd-5.15.25-rockchip64
uInitrd-5.10.60-rockchip64:     u-boot legacy uImage, uInitrd, Linux/ARM 64-bit, RAMDisk Image (gzip), 16449299 bytes, Wed Mar 16 22:04:17 2022, Load Address: 00000000, Entry Point: 00000000, Header CRC: 0XE8719F5C, Data CRC: 0XD84B7EFC
uInitrd-5.15.25-rockchip64:     u-boot legacy uImage, uInitrd, Linux/ARM 64-bit, RAMDisk Image (gzip), 15658552 bytes, Thu Mar 17 20:47:32 2022, Load Address: 00000000, Entry Point: 00000000, Header CRC: 0XA3E0EE71, Data CRC: 0XDFFB8870
vmlinuz-5.15.25-rockchip64:     Linux kernel ARM64 boot executable Image, little-endian, 4K pages

```

Inside `Void/boot`:

```
config-5.15.30_1:        Linux make config build file, ASCII text
dtbs:                    directory
initramfs-5.15.30_1.img: regular file, no read permission
vmlinux-5.15.30_1:       Linux kernel ARM64 boot executable Image, little-endian, 4K pages
```

**Comment by Anachron** [comment]
After digging a bit more I believe the issue is that https://github.com/void-linux/void-mklive/blob/master/mkplatformfs.sh.in#L180 is not trigger for my `platformfs` build, even though I have the tools but not the files:

```sh
ls -lisah usr/bin/{mkimage,dracut}
919096  88K -rwxr-xr-x 1 root root  85K 14. Jul 2021  usr/bin/dracut
919116 300K -rwxr-xr-x 1 root root 299K 21. Jan 01:10 usr/bin/mkimage
```

```sh
LC_ALL=C ls -lisah boot/*nitrd
ls: cannot access 'boot/*nitrd': No such file or directory
```

Probably because of https://github.com/void-linux/void-mklive/blob/master/mkplatformfs.sh.in#L182, I build for `aarch64` but it checks for `arm` in `XBPS_TARGET_ARCH`.

**Comment by zen0bit** [comment]
Little 'offtopic' question:
Can be used for pinephone pro or at least as base + some phone stuff?

I am not experienced in this..

but wanna void also on ppp and wondering where to start...

**Comment by Anachron** [comment]
> @Anachron your comments mention the master branch of [void-mklive](https://github.com/void-linux/void-mklive). I used the files from [void-linux/void-mklive#221](https://github.com/void-linux/void-mklive/pull/221) to create the image, rootfs and platformfs files. Perhaps that is why the scripts don't work for you.
> 
> Also note that I made some changes to the mkimage.sh.in script described [here](https://github.com/void-linux/void-mklive/pull/221#discussion_r735122038).

I'm confused by your reply. Earlier you said you were using the `mkplatformfs.sh` but now you're saying I should modify the `mklive.sh` script?

Which commands did you run exactly to make `RockPro64` work? And which changes did you make to the build process?

I did finally get my sdcard to  boot, even though it complains about not being able to find anything after running the following commands (I've used the precompiled blobs from rockpro64-uboot):

https://github.com/u-boot/u-boot/blob/master/doc/README.rockchip#L457

I'm trying the `mklive.sh` way again in the meantime hoping to get something working from this angle.

**Comment by CameronNemo** [comment]
>it builds the initrd now

I do not think you want an initrd or uinitrd file. The `initramfs-5.15.30_1.img` file should be enough if you are using u-boot-menu.

>Also note that I made some changes to the mkimage.sh.in script described https://github.com/void-linux/void-mklive/pull/221#discussion_r735122038.

Someone should probably make a new PR for the mklive repo that includes those changes. That PR's submitter does not seem particularly active. Is that something you can take care of, @zen0bit? I do not have the hardware but I can review based on what I know from the PBP.

>Even can't get it to boot using mkimage.sh

Yeah that is expected. mkimage will just be a plain old aarch64 rootfs. It will not have a partition table set up, or the U-boot blobs written to the correct offsets. You feed the output of mkimage to mkplatformfs typically.

**Comment by Anachron** [comment]
@CameronNemo in case you missed it, I can boot my RockPro64 on Void Linux now. 
I'll test it thoroughly and reply here if I find something important not working.

**Comment by tsndqst** [comment]
Sorry for not replying sooner, I have been busy.

First, a response to @Anachron's questions.
>I'm confused by your reply. Earlier you said you were using the mkplatformfs.sh but now you're saying I should modify the mkimage.sh script?

There are outstanding changes in https://github.com/void-linux/void-mklive/pull/221 that allow the `void-mklive` tools to build a platform-specific image for RockPro64s.  There was code missing from the original author's change.  The mkimage.sh script is used to create an image based on the tar file created by mkplatformfs.sh.

>Which commands did you run exactly to make RockPro64 work? And which changes did you make to the build process?

It has been a while since I created a new image but this is roughly what I did:
1. I used `mkrootfs.sh` to make an aarch64 root fs tar file.
2. I used `mkplatformfs.sh` to make a platform fs tar file specific to the rockpro64
3. I used `mkimage.sh` to make an image suitable to be written to eMMC or an SD card (I have only used eMMC but it should work on SD card too)
 
>Someone should probably make a new PR for the mklive repo that includes those changes. That PR's submitter does not seem particularly active. Is that something you can take care of, @tsndqst ? I do not have the hardware but I can review based on what I know from the PBP.

@CameronNemo I can make a new PR for `void-mklive`.

File: srcpkgs/rockpro64-base/template [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Template file for 'rockpro64-base'
+pkgname=rockpro64-base
+version=1
+revision=1
+archs="aarch64*"
+build_style=meta
+depends="rockpro64-uboot u-boot-menu dracut"
```
Comment (tsndqst) [inline srcpkgs/rockpro64-base/template:7]: The other base templates such as pinephone-base and pinephonepro-base include a kernel in the depends.  Even though the rockpro64 doesn't require a custom kernel we should include `linux` as a dependency so images created with `rockpro64-base` will automatically have a kernel included.
```suggestion
depends="linux rockpro64-uboot u-boot-menu dracut"
```

Comment (CameronNemo) [reply srcpkgs/rockpro64-base/template:7]: I think this is a good idea.

Comment (Anachron) [reply srcpkgs/rockpro64-base/template:7]: Yep, this did bite me earlier trying to figure out why there is no kernel installed in my image. Took me some time to realize this was just left out from the base-pkg.

Comment (tsndqst) [reply srcpkgs/rockpro64-base/template:7]: I think I'll have some time this weekend so if @tiemenwerkman doesn't respond to the above suggestion I can make a new PR.

**Comment by Anachron** [comment]
@tsndqst could you please open up your own PR, OP seems inactive.

**Comment by Anachron** [comment]
We have a new PR. This can probably be closed.

**Pull request closed without merging**